Manage excluded_products in ranges dashboard

gwaidacher opened the pull request:
Added tabs to range - products in dashboard, first tab contains now existing form for managing included products, in the second tab the excluded products can be managed, with a separate textfield for upc input and a second fileupload form.

see https://github.com/django-oscar/django-oscar/issues/4055

gwaidacher commented:
Added missing tests. where does the "1 file has unexpected coverage changes not visible in diff " error come from?

joeyjurjens reviewed (commented):
Hi @gwaidacher , thank you very much for your PR. It's quite a big one and we really appreciate it that you're taking the time for it!

I do have some review comments though, if you have any questions or comments to mine, just let me know! :)

gwaidacher commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -130,41 +148,86 @@ def remove_selected_products(self, request, products):
+        self.handle_query_products(request, range, form)
+        self.handle_file_products(request, range, form, included=False)
+        return redirect(reverse('dashboard:range-products-excluded',
+                                kwargs={'pk': range.pk, 'excluded': 1}))
```
It's used to open the 2nd Tab with the excluded products when the page is loaded.

gwaidacher commented on src/oscar/views/generic.py:
```diff
@@ -57,8 +57,8 @@ def get_checkbox_object_name(self):
 
-    def get_success_url(self, request):
-        return safe_referrer(request, '.')
+    # def get_success_url(self, request):
```
It is not used any more which lead to the "1 file has unexpected coverage changes not visible in diff " Error in codecov report.

gwaidacher commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
     paginate_by = settings.OSCAR_DASHBOARD_ITEMS_PER_PAGE
 
+    def get(self, request, **kwargs):
+        self.excluded = 'excluded' in kwargs
```
its used to switch to the 2nd Tab with the excluded_products (see ctx['active_tab'] in get_context_data() and on form.in_valid() in add_excluded_products() )

gwaidacher commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -130,41 +148,86 @@ def remove_selected_products(self, request, products):
+    def add_excluded_products(self, request):
+        range = self.get_range()
+        form = self.form_class(range, request.POST, request.FILES)
+        form.included = False
```
I have tried it before with the hidden boolean input and passing it via argument through the __init__ method but I received form errors with it which I could not resolve. My impatience made me solve it this way, though I know it's not the django way...

gwaidacher commented on src/oscar/apps/offer/abstract_models.py:
```diff
@@ -1042,6 +1042,8 @@ class AbstractRangeProductFileUpload(models.Model):
         verbose_name=_("Uploaded By"))
     date_uploaded = models.DateTimeField(_("Date Uploaded"), auto_now_add=True, db_index=True)
+    # for managing excluded_products
+    included = models.BooleanField(default=True)
```
Hi! I restructured this PR fitting your suggestions, please have a look!

gwaidacher commented:
* changed field "included to "upload_type" in RangeProductFileUpload
* changed the "excluded" kwarg in redirect url to "active_tab"
* passed the "excluded" argument in RangeProductForm via __init__() method

gwaidacher commented on src/oscar/apps/offer/abstract_models.py:
```diff
@@ -1090,9 +1092,13 @@ def was_processing_successful(self):
         """
         all_ids = set(self.extract_ids(file_obj))
-        products = self.range.all_products()
+        if self.included:
```
Done.

gwaidacher replied:
done.

gwaidacher commented:
> Hi @gwaidacher , thank you very much for your PR. It's quite a big one and we really appreciate it that you're taking the time for it!
> 
> I do have some review comments though, if you have any questions or comments to mine, just let me know! :)

PR updated, please have a look

viggo-devries commented on src/oscar/apps/dashboard/ranges/forms.py:
```diff
@@ -31,8 +31,9 @@ class RangeProductForm(forms.Form):
         help_text=_('Either comma-separated, or one identifier per line'))
 
-    def __init__(self, range, *args, **kwargs):
+    def __init__(self, range, excluded, *args, **kwargs):
```
please make this a kwarts, excluded=True/False, otherwise it's really hard to know in the view what this does.

gwaidacher replied:
done.

viggo-devries replied:
Yes, and then you can remove it from the __init__ args.

gwaidacher replied:
done.

joeyjurjens commented on src/oscar/apps/dashboard/ranges/apps.py:
```diff
@@ -27,6 +27,8 @@ def get_urls(self):
             path('<int:pk>/', self.update_view.as_view(), name='range-update'),
             path('<int:pk>/delete/', self.delete_view.as_view(), name='range-delete'),
             path('<int:pk>/products/', self.products_view.as_view(), name='range-products'),
+            path('<int:pk>/products/<int:active_tab>/', self.products_view.as_view(),
```
I still don't think this is necessary. If you want to remember what the active tab was before submiting the form, I suggest using a new HiddenFormField which you initially set on the forms.

joeyjurjens commented on src/oscar/apps/dashboard/ranges/forms.py:
```diff
@@ -33,6 +33,7 @@ class RangeProductForm(forms.Form):
 
     def __init__(self, range, *args, **kwargs):
         self.range = range
+        self.excluded = kwargs.pop('excluded', False)
```
If you add a new hidden input field, which describes what kind of form it is, you can just remove this.
Inside the clean method you could just access the hidden form field and do logic based on that.

```python
upload_type = forms.CharField(widget=forms.HiddenInput())
```

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
     form_class = RangeProductForm
     paginate_by = settings.OSCAR_DASHBOARD_ITEMS_PER_PAGE
 
+    def get(self, request, **kwargs):
```
With the hidden form field, you can also remove this

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -96,14 +95,24 @@ class RangeProductListView(BulkEditMixin, ListView):
+        return super().get(request, **kwargs)
+
     def post(self, request, *args, **kwargs):
+        self.active_tab = kwargs.get('active_tab', 0)
```
With the hidden form field, you can also remove this

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -121,6 +130,13 @@ def get_context_data(self, **kwargs):
         ctx['range'] = range
         if 'form' not in ctx:
             ctx['form'] = self.form_class(range)
```
Here, you would set the intiial value of the upload type. Say you named the hidden field: upload_type, you can do this:

```python
ctx['form'] = self.form_class(range, initial={"upload_type": INCLUDED_PRODUCTS_TYPE})
```

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -121,6 +130,13 @@ def get_context_data(self, **kwargs):
         if 'form' not in ctx:
             ctx['form'] = self.form_class(range)
+        if 'form_excluded' not in ctx:
+            ctx['form_excluded'] = self.form_class(range, excluded=True)
```
And here this would be:

```python
ctx['form_excluded'] = self.form_class(range, initial={"upload_type": EXCLUDED_PRODUCTS_TYPE})
```

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -143,28 +162,68 @@ def add_products(self, request):
+
+    def add_excluded_products(self, request):
+        range = self.get_range()
+        form = self.form_class(
```
Here you can remove the `excluded=True` as you have access to the upload type:
form.cleaned_data["upload_type"] and thus you also know what the active tab will be.

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -174,11 +233,12 @@ def handle_query_products(self, request, range, form):
         self.check_imported_products_sku_duplicates(request, products)
 
-    def handle_file_products(self, request, range, form):
+    def handle_file_products(self, request, range, form, included=True):
```
Again, you could remove included=True here because you can access the type from the `form.cleaned_data["upload_type"]`

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -190,12 +250,17 @@ def handle_file_products(self, request, range, form):
         self.check_imported_products_sku_duplicates(request, products)
 
-    def create_upload_object(self, request, range, f):
+    def create_upload_object(self, request, range, f, included=True):
```
Same as above, you can remove included here, because you have the upload type in `f.cleaned_data["upload_type"]`

joeyjurjens commented on src/oscar/views/generic.py:
```diff
@@ -57,8 +57,8 @@ def get_checkbox_object_name(self):
 
-    def get_success_url(self, request):
-        return safe_referrer(request, '.')
+    # def get_success_url(self, request):
```
I still don't see how this method has anything to do with this PR? Can't you just keep it like it was?

gwaidacher replied:
There was a codecov error concerning this function, so I commented it out to have all checks passed. It's not needed any more by ranges.views.py

joeyjurjens commented on tests/functional/dashboard/test_range.py:
```diff
@@ -20,7 +20,8 @@ def tearDown(self):
 
     def submit_form(self, data):
-        return forms.RangeProductForm(self.range, data)
+        form = forms.RangeProductForm(self.range, data)
```
Is this change really necessary?

joeyjurjens reviewed (changes requested):
Hi @gwaidacher , I still have some feedback to keep it a bit more clean. Thanks again for your time and effort.

gwaidacher commented on src/oscar/apps/dashboard/ranges/forms.py:
```diff
@@ -30,9 +31,15 @@ class RangeProductForm(forms.Form):
     def __init__(self, range, *args, **kwargs):
         self.range = range
+        # make upload_type accessible in clean_query
+        if kwargs.get("initial"):
```
I cannot access cleaned_data['upload_type'] in clean_query() - the solution to avoid this would be to merge clean_query() into clean()

joeyjurjens commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -144,27 +166,73 @@ def add_products(self, request):
+        self.handle_file_products(request, range, form)
+        return redirect(
+            reverse('dashboard:range-products', kwargs={'pk': range.pk})
+            + '?upload_type=excluded'
```
I don't think this is really necessary. Once the upload was successful, we can just redirect them as is, maybe with a success message stating it was successful. The GET parameter is not really needed in my opinion.

gwaidacher replied:
It's used to redirected to the page having the second tab with excluded products active.

joeyjurjens reviewed (changes requested):
Few small things left, the rest looks great!

gwaidacher commented on src/oscar/apps/dashboard/ranges/forms.py:
```diff
@@ -40,43 +42,47 @@ def clean(self):
-            raise forms.ValidationError(
-                _("The products with SKUs or UPCs matching %s are already in"
-                  " this range") % (', '.join(ids)))
+        raw = clean_data['query']
```
It was not possible to add the changes in clean_query, because I could not access `clean_data['upload_type']` there. Otherwise I would have to call super.clean() twice. Any ideas?

gwaidacher replied:
ah, misunderstood... fixed.

viggo-devries commented on src/oscar/templates/oscar/dashboard/ranges/range_product_list.html:
```diff
@@ -30,114 +30,248 @@
+                  <div class="float-right">
+                    <input type="hidden" name="action" value="remove_excluded_products" />
+                    <button type="submit" class="btn btn-secondary" data-loading-text="{% trans 'Removing...' %}">
+                      {% trans "Remove excluded products" %}
```
Please change this to:

Remove selected excluded products

gwaidacher replied:
fixed

viggo-devries commented on src/oscar/apps/dashboard/ranges/views.py:
```diff
@@ -144,27 +166,73 @@ def add_products(self, request):
-
+            if form.cleaned_data["upload_type"] == \
+                    RangeProductFileUpload.EXCLUDED_PRODUCTS_TYPE:
+                action = _('excluded from this range')
```
please add a test for this

viggo-devries commented:
In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.

gwaidacher commented:
> In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.

Hi! There are already 2 tests in test_range.RangeProductViewTest: test_upload_excluded_file_with_skus for the upload-file form and test_dupe_excluded_skus_warning - if this is insufficient, please specify which function should be tested.

viggo-devries commented:
> > In general this is missing some tests, great that the existing tests still pass but please add some specific tests for excluding the products.
> 
> Hi! There are already 2 tests in test_range.RangeProductViewTest: test_upload_excluded_file_with_skus for the upload-file form and test_dupe_excluded_skus_warning - if this is insufficient, please specify which function should be tested.

Yes I saw but the fact that your tests missed this entire functionality: https://github.com/django-oscar/django-oscar/pull/4073/commits/f97053a741a3a65d35b8d45d563c46480c500325 is not good.

I would like to see tests for every action that can be performed and also the result of that action.
You should be able to come up with way more than the 2 actions you have tested now. like:
test_add_single_product_to_excluded_range
test_add_multipe_products_to_excluded_range
remove_single_product_from_excluded_range
remove_multiple_products_from_excluded_range
etc.etc.

joeyjurjens commented:
@gwaidacher Hey man. I have written the missing tests, so you don't have to do that anymore.

gwaidacher commented:
> @gwaidacher Hey man. I have written the missing tests, so you don't have to do that anymore.

Thank you very much :D
